divi/src/test/kernel_tests.cpp

```diff
@@ -0,0 +1,174 @@
+
+protected:
+
+  CScript scriptVault;
+  CScript scriptNonVault;
+
+  std::vector<COutPoint> vaultCoins;
+  std::vector<COutPoint> nonVaultCoins;
+
+  CheckCoinstakeForVaultsTestFixture()
+    : coins(&coinsDummy), rewards(CENT, 0, 0, 0, 0, 0)
+  {
+    const std::vector<unsigned char> key1(20, 'x');
+    const std::vector<unsigned char> key2(20, 'y');
+
+    scriptVault = CreateStakingVaultScript(key1, key2);
+    scriptNonVault = CScript() << OP_TRUE;
+
+    BOOST_CHECK(IsStakingVaultScript(scriptVault));
+    BOOST_CHECK(!IsStakingVaultScript(scriptNonVault));
+
+    CMutableTransaction dummyTxVault;
+    CMutableTransaction dummyTxNonVault;
+    dummyTxVault.vout.resize(2);
+    dummyTxNonVault.vout.resize(dummyTxVault.vout.size());
+    for (unsigned i = 0; i < dummyTxVault.vout.size(); ++i) {
+      dummyTxVault.vout[i].nValue = COIN;
+      dummyTxVault.vout[i].scriptPubKey = scriptVault;
+      dummyTxNonVault.vout[i].nValue = COIN;
+      dummyTxNonVault.vout[i].scriptPubKey = scriptNonVault;
+    }
+
+    coins.ModifyCoins(dummyTxVault.GetHash())->FromTx(dummyTxVault, 0);
+    coins.ModifyCoins(dummyTxNonVault.GetHash())->FromTx(dummyTxNonVault, 0);
+
+    for (unsigned i = 0; i < dummyTxVault.vout.size(); ++i) {
+      vaultCoins.emplace_back(dummyTxVault.GetHash(), i);
+      nonVaultCoins.emplace_back(dummyTxNonVault.GetHash(), i);
+    }
+  }
+
+  bool RunCheck(const CMutableTransaction& mtx)
+  {
+    return CheckCoinstakeForVaults(CTransaction(mtx), rewards, coins);
+  }
+
+};
+
+BOOST_FIXTURE_TEST_SUITE(CheckCoinstakeForVaults_tests, CheckCoinstakeForVaultsTestFixture)
+
+BOOST_AUTO_TEST_CASE(NoCoinstake)
+{
+  CMutableTransaction mtx;
+  mtx.vin.push_back(CTxIn(nonVaultCoins[0]));
+  mtx.vout.push_back(CTxOut(1, scriptNonVault));
+  mtx.vout.push_back(CTxOut(COIN + CENT, scriptNonVault));
+
+  BOOST_CHECK(!CTransaction(mtx).IsCoinStake());
+  BOOST_CHECK(!RunCheck(mtx));
+  
+  mtx.vout[0].SetEmpty();
+  BOOST_CHECK(CTransaction(mtx).IsCoinStake());
+  BOOST_CHECK(RunCheck(mtx));
+}
+
+BOOST_AUTO_TEST_CASE(NoVaultUsed)
+{
+  CMutableTransaction mtx;
+  mtx.vin.push_back(CTxIn(nonVaultCoins[0]));
+  mtx.vin.push_back(CTxIn(nonVaultCoins[1]));
+
+  mtx.vout.push_back(CTxOut());
+  mtx.vout[0].SetEmpty();
+  mtx.vout.push_back(CTxOut(COIN / 2, scriptNonVault));
+  mtx.vout.push_back(CTxOut(COIN / 2, scriptNonVault));
+
+  BOOST_CHECK(CTransaction(mtx).IsCoinStake());
+  BOOST_CHECK(RunCheck(mtx));
+}
+
+BOOST_AUTO_TEST_CASE(CorrectVaultPayment)
+{
+  CMutableTransaction mtx;
+  mtx.vin.push_back(CTxIn(vaultCoins[0]));
+  mtx.vin.push_back(CTxIn(vaultCoins[1]));
+
+  mtx.vout.push_back(CTxOut());
+  mtx.vout[0].SetEmpty();
+  mtx.vout.push_back(CTxOut(2 * COIN + CENT, scriptVault));
+  mtx.vout.push_back(CTxOut(CENT, scriptNonVault));
+  BOOST_CHECK(CTransaction(mtx).IsCoinStake());
+  BOOST_CHECK(RunCheck(mtx));
+
+  mtx.vout[1].nValue += 1;
+  BOOST_CHECK(RunCheck(mtx));
+}
+
+BOOST_AUTO_TEST_CASE(VaultPaidToWrongScript)
```

Comment:
Test rename suggestion: willNotAllowPaymentToNonVault

Importantly, I don't see a test for paying to *another* different vault :P